core/corehttp/gateway_handler.go

```diff
@@ -174,15 +174,7 @@ func (i *gatewayHandler) getHandler(w http.ResponseWriter, r *http.Request) {
 
 	nd, p, err := i.ResolvePath(ctx, urlPath)
 	if err != nil {
-		if err == routing.ErrNotFound {
-			w.WriteHeader(http.StatusNotFound)
-		} else if err == context.DeadlineExceeded {
-			w.WriteHeader(http.StatusRequestTimeout)
-		} else {
-			w.WriteHeader(http.StatusBadRequest)
-		}
-
-		w.Write([]byte(err.Error()))
```

Comment:
I think this change is incorrect, because:
- when routing returns `ErrNotFound` we want to say 404. (it means routing exhausted the search)
- when routing times out (didnt finish searching) we want 408 (`http.StatusRequestTimeout`)

I think the problem in #924 is that the error returned is -- as you point out `path.ErrNoLink`, which we should special-case to 404. We likely have more errors being returned to special case. it's likely that we just dont cover them all yet.

> well, I get that error message, but it's sent by the server as a 404

this is weird to me-- is it somehow sending the 404 before it lands on `w.WriteHeader(http.StatusBadRequest)` here?